File: projects/meta-mcp-server/jetbrains-plugin/src/main/kotlin/com/adobe/metamcp/services/RepoDetector.kt
package com.adobe.metamcp.services

import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.components.Service
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.project.ProjectManager
import java.io.File
import java.util.concurrent.TimeUnit

private val isMac = System.getProperty("os.name").lowercase().contains("mac")
private val isWindows = System.getProperty("os.name").lowercase().contains("win")
private val homeDir = System.getProperty("user.home")

/**
 * Repository Detector - Auto-detects local repository clones
 * Port of extension/src/services/RepoDetector.ts
 */
@Service(Service.Level.APP)
class RepoDetector {
    private val log = Logger.getInstance(RepoDetector::class.java)

    companion object {
        fun getInstance(): RepoDetector =
            ApplicationManager.getApplication().getService(RepoDetector::class.java)
    }

    /**
     * Auto-detect the location of a repository (e.g., adobe-mcp-servers)
     * Uses multiple strategies: IntelliJ open projects, Spotlight (macOS), find command
     */
    fun findRepository(repoName: String): String? {
        // 1. Check currently open projects in IntelliJ
        val projectManager = ProjectManager.getInstance()
        for (project in projectManager.openProjects) {
            val basePath = project.basePath
            if (basePath != null) {
                if (basePath.contains(repoName)) {
                    return basePath
                }
                // Also check if it's a subfolder
                val subPath = File(basePath, repoName)
                if (subPath.exists()) {
                    return subPath.absolutePath
                }
            }
        }

        // 2. Check downloaded repos directory
        val downloadedPath = RepoDownloader.getInstance().getRepositoryPath(repoName)
        if (downloadedPath != null) {
            return downloadedPath
        }

        // 3. macOS Spotlight search (instant, indexed)
        if (isMac) {
            findWithSpotlight(repoName)?.let { return it }
        }

        // 4. Fast find with depth limit (works on all platforms)
        findWithSystemSearch(repoName)?.let { return it }

        return null
    }

    /**
     * Use macOS Spotlight for fast indexed search
     */
    private fun findWithSpotlight(repoName: String): String? {
        return try {
            val process = ProcessBuilder("mdfind", "-name", repoName, "-onlyin", homeDir)
                .redirectErrorStream(true)
                .start()

            val result = process.inputStream.bufferedReader().readText().trim()
            process.waitFor(3, TimeUnit.SECONDS)

            result.lines()
                .firstOrNull { it.endsWith("/$repoName") && File(it).exists() }
        } catch (e: Exception) {
            log.warn("Spotlight search failed: ${e.message}")
            null
        }
    }

    /**
     * Use system find/dir command for directory search
     */
    private fun findWithSystemSearch(repoName: String): String? {
        return try {
            val command = if (isWindows) {
                listOf("cmd", "/c", "dir", "/s", "/b", "/ad", homeDir, "|", "findstr", repoName)
            } else {
                listOf("find", homeDir, "-maxdepth", "5", "-type", "d", "-name", repoName)
            }

            val process = ProcessBuilder(command)
                .redirectErrorStream(true)
                .start()

            val result = process.inputStream.bufferedReader().readText().trim()
            process.waitFor(10, TimeUnit.SECONDS)

            result.lines()
                .firstOrNull { it.isNotEmpty() && File(it).exists() }
        } catch (e: Exception) {
            log.warn("Find command failed: ${e.message}")
            null
        }
    }

    /**
     * Validate that a path contains the expected package
     */
    fun validatePackagePath(repoPath: String, packagePath: String): Boolean {
        val fullPath = File(repoPath, packagePath)
        return fullPath.exists()
    }

    /**
     * Find a local server within a repository
     * Checks common locations: src/{id}, packages/{id}
     */
    fun findLocalServer(repoPath: String, serverId: String): String? {
        val candidates = listOf(
            File(repoPath, "src/$serverId"),
            File(repoPath, "packages/$serverId"),
            File(repoPath, serverId)
        )

        return candidates.firstOrNull { it.exists() && it.isDirectory }?.absolutePath
    }
}

File: extension/src/services/RepoDetector.ts
import * as vscode from 'vscode';
import * as fs from 'fs';
import * as path from 'path';
import { execSync } from 'child_process';

/**
 * Auto-detect the location of a repository (e.g., adobe-mcp-servers)
 * Uses multiple strategies: VS Code workspace, Spotlight (macOS), find command
 */
export async function findRepository(repoName: string): Promise<string | null> {
    // 1. Check VS Code workspace folders first (most reliable if user has it open)
    const workspaceFolders = vscode.workspace.workspaceFolders;
    if (workspaceFolders) {
        for (const folder of workspaceFolders) {
            if (folder.uri.fsPath.includes(repoName)) {
                return folder.uri.fsPath;
            }
            // Also check if it's a subfolder
            const subPath = path.join(folder.uri.fsPath, repoName);
            if (fs.existsSync(subPath)) {
                return subPath;
            }
        }
    }
    
    // 2. macOS Spotlight search (instant, indexed)
    if (process.platform === 'darwin') {
        try {
            const result = execSync(
                `mdfind -name "${repoName}" -onlyin ~ 2>/dev/null | grep -E "/${repoName}$" | head -1`,
                { encoding: 'utf-8', timeout: 3000 }
            ).trim();
            if (result && fs.existsSync(result)) {
                return result;
            }
        } catch {
            // Spotlight search failed, continue to fallback
        }
    }
    
    // 3. Fast find with depth limit (works on all platforms)
    const homeDir = process.env.HOME || process.env.USERPROFILE || '~';
    try {
        const result = execSync(
            `find "${homeDir}" -maxdepth 5 -type d -name "${repoName}" 2>/dev/null | head -1`,
            { encoding: 'utf-8', timeout: 10000 }
        ).trim();
        if (result && fs.existsSync(result)) {
            return result;
        }
    } catch {
        // Find command failed
    }
    
    return null;
}

/**
 * Prompt user to select a repository folder
 */
export async function promptForRepository(repoName: string): Promise<string | null> {
    const selected = await vscode.window.showOpenDialog({
        canSelectFolders: true,
        canSelectFiles: false,
        canSelectMany: false,
        title: `Select ${repoName} repository folder`,
        openLabel: 'Select Repository'
    });
    
    if (!selected || selected.length === 0) {
        return null;
    }
    
    return selected[0].fsPath;
}

/**
 * Find repository with fallback to user prompt
 */
export async function findOrPromptForRepository(repoName: string): Promise<string | null> {
    // Try auto-detection first
    const autoDetected = await findRepository(repoName);
    
    if (autoDetected) {
        // Confirm with user
        const useDetected = await vscode.window.showInformationMessage(
            `Found ${repoName} at: ${autoDetected}`,
            'Use This',
            'Choose Different'
        );
        
        if (useDetected === 'Use This') {
            return autoDetected;
        }
    }
    
    // Fall back to manual selection
    return promptForRepository(repoName);
}

/**
 * Validate that a path contains the expected package
 */
export function validatePackagePath(repoPath: string, packagePath: string): boolean {
    const fullPath = path.join(repoPath, packagePath);
    return fs.existsSync(fullPath);
}
